### PR TITLE
Parallelize compile.erlang

### LINKED ISSUE #10690: Parallelize Erlang compilation
Today we compile serially due to simply walking over the toposorted list of parse transform and behaviour compile time dependencies. We could do this in parallel by only compiling the dependencies serially and then all dependents in parallel.

EDIT1: We can also do the parsing with `:epp.parse_file/3` in parallel.
EDIT2: Another potential optimizations is to pass the forms from `:epp.parse_file/3` to `:compile.forms/2` instead of `:compile.file/2` to avoid parsing files twice.

### PR DESCRIPTION
**Pull request by ericmj**
Improves the compilation time of [erlang-jose](https://github.com/potatosalad/erlang-jose) from 3.7s to 1.6s and increases CPU utilization from 125% to 400% on my machine.

Closes #10690.

### CODE REVIEW

File: lib/mix/lib/mix/tasks/compile.erlang.ex [hunk 1]
```diff
@@ -74,7 +74,15 @@ defmodule Mix.Tasks.Compile.Erlang do
     end
 
     erlc_options =
-      erlc_options ++ [:debug_info, :return, :report, outdir: compile_path, i: include_path]
+      erlc_options ++
+        [
+          :debug_info,
+          :return,
+          :report,
+          :no_spawn_compiler_process,
```
Comment (ericmj) [inline lib/mix/lib/mix/tasks/compile.erlang.ex:82]: @josevalim is this the option you meant?

File: lib/mix/lib/mix/compilers/erlang.ex [hunk 2]
```diff
@@ -119,11 +119,21 @@ defmodule Mix.Compilers.Erlang do
...
+      parallel_sources = opts[:parallel] || MapSet.new()
+
+      {parallel, serial} =
+        Enum.split_with(stale, fn {source, _target} -> source in parallel_sources end)
+
+      serial_results = Enum.map(serial, &do_compile(&1, callback, timestamp, verbose))
+
+      parallel_results =
+        parallel
+        |> Task.async_stream(&do_compile(&1, callback, timestamp, verbose))
```
Comment (lukaszsamson) [inline lib/mix/lib/mix/compilers/erlang.ex:131]: Shouldn't timeouts be handled here?

Comment (ericmj) [reply lib/mix/lib/mix/compilers/erlang.ex:131]: Yes, we need infinite timeout.

File: lib/mix/lib/mix/compilers/erlang.ex [hunk 3]
```diff
@@ -119,11 +119,21 @@ defmodule Mix.Compilers.Erlang do
       # and what not).
       Code.prepend_path(Mix.Project.compile_path())
 
+      parallel_sources = opts[:parallel] || MapSet.new()
+
+      {parallel, serial} =
+        Enum.split_with(stale, fn {source, _target} -> source in parallel_sources end)
```
Comment (josevalim) [inline lib/mix/lib/mix/compilers/erlang.ex:125]:
```suggestion
      {parallel, serial} =
        case opts[:parallel] || false do
          true -> {stale, []}
          false -> {[], stale}
          parallel -> Enum.split_with(stale, fn {source, _target} -> source in parallel end)
        end
```

File: lib/mix/lib/mix/tasks/compile.erlang.ex (hunk 1 above)
Comment (josevalim) [reply lib/mix/lib/mix/tasks/compile.erlang.ex:82]: Yes!

Comment (josevalim) [reply lib/mix/lib/mix/tasks/compile.erlang.ex:82]: But you want to set it only inside the parallel compilation... If it is not easy to do that, then we should remove this option.

File: lib/mix/lib/mix/tasks/compile.erlang.ex [hunk 4]
```diff
@@ -84,11 +92,10 @@ defmodule Mix.Tasks.Compile.Erlang do
...
 
-    tuples =
-      files
-      |> scan_sources(include_path, source_paths)
-      |> sort_dependencies()
-      |> Enum.map(&annotate_target(&1, compile_path, opts[:force]))
+    erls = scan_sources(files, include_path, source_paths)
+    tuples = Enum.map(erls, &annotate_target(&1, compile_path, opts[:force]))
+    parallel = Enum.map(find_parallel(erls), & &1.file)
+    opts = opts ++ [parallel: MapSet.new(parallel)]
```
Comment (josevalim) [inline lib/mix/lib/mix/tasks/compile.erlang.ex:98]:
```suggestion
    opts = [parallel: MapSet.new(parallel)] ++ opts
```

File: lib/mix/lib/mix/tasks/compile.erlang.ex [hunk 5]
```diff
@@ -84,11 +92,10 @@ defmodule Mix.Tasks.Compile.Erlang do
 
     compile_path = Path.relative_to(compile_path, File.cwd!())
 
-    tuples =
-      files
-      |> scan_sources(include_path, source_paths)
-      |> sort_dependencies()
-      |> Enum.map(&annotate_target(&1, compile_path, opts[:force]))
+    erls = scan_sources(files, include_path, source_paths)
+    tuples = Enum.map(erls, &annotate_target(&1, compile_path, opts[:force]))
```
Comment (josevalim) [inline lib/mix/lib/mix/tasks/compile.erlang.ex:96]: Shouldn't we do the target annotation in parallel too? (i.e. inside scan_sources).

File: lib/mix/lib/mix/tasks/compile.erlang.ex [hunk 6]
```diff
@@ -126,10 +133,16 @@ defmodule Mix.Tasks.Compile.Erlang do
 
   defp scan_sources(files, include_path, source_paths) do
     include_paths = [include_path | source_paths]
-    Enum.reduce(files, [], &scan_source(&2, &1, include_paths)) |> Enum.reverse()
+
+    files
+    |> Task.async_stream(&scan_source(&1, include_paths))
```
Comment (josevalim) [inline lib/mix/lib/mix/tasks/compile.erlang.ex:138]: Don't forget the timeout. We might be fine with ordered: false too?

File: lib/mix/lib/mix/tasks/compile.erlang.ex (hunk 1 above)
Comment (ericmj) [reply lib/mix/lib/mix/tasks/compile.erlang.ex:82]: There is no nice way of doing it unfortunately. How about we set the option always and spawn and wait for processes serially for non-parallel compilation?

Comment (josevalim) [reply lib/mix/lib/mix/tasks/compile.erlang.ex:82]: Given we are compiling from a file and not from AST, setting this option is not really relevant. So it is fine to not include it.

**Review by josevalim — APPROVED** [review]

**Pull request merged**